chain/syncer.go

```diff
@@ -423,17 +340,53 @@ func (syncer *Syncer) HandleNewTipset(ctx context.Context, tsKey types.TipSetKey
 	defer syncer.mu.Unlock()
 
 	// If the store already has this tipset then the syncer is finished.
-	if _, err := syncer.chainStore.GetTipSet(tsKey); err == nil {
+	if syncer.chainStore.HasTipSetAndState(ctx, tsKey.String()) {
 		return nil
 	}
 
-	// Walk the chain given by the input blocks back to a known tipset in
-	// the store. This is the only code that may go to the network to
-	// resolve cids to blocks.
-	chain, err := syncer.collectChain(ctx, tsKey)
-	if err != nil {
-		return err
+	syncDoneCb := func(t types.TipSet) (bool, error) {
+		return syncer.chainStore.HasTipSetAndState(ctx, t.String()), nil
```

Comment:
I think this can be `HasTipSetAndState(t.Parent.String())`, assuming the fetcher still returns the tipset for which the predicate returns `true`. This has the effect of stopping the fetch one tip sooner.